File: abstract_factory/kotlin/Factory.kt
interface Silla{
    fun description(): String
}

interface Sofa{
    fun description(): String
}

interface Mesilla{
    fun description(): String
}


// Tenemos diferente variadades de estos productos Moderno y Victoriano 


//  Juego de estilo Victoriano
class SillaVictoriana : Silla{

    override fun description()= "silla victoriano"

}

class SofaVictoriano : Sofa{

    override fun description(): String = "sofa victoriano"

}

class MesillaVictoriana : Mesilla{

    override fun description(): String = "mesilla victoriano"

}


//Juego de estilo Moderno

class SillaModerna : Silla{

    override fun description()= "silla moderna"

}

class SofaModerno : Sofa{

    override fun description(): String = "sofa moderno"

}

class MesillaModerna : Mesilla{

    override fun description(): String = "mesilla moderna"

}

abstract class Factory {
    
    abstract fun crearSilla(): Silla

    abstract fun crearSofa(): Sofa

    abstract fun crearMesilla(): Mesilla


    fun combinarMuebles(): String =
        "Juego de muebles Silla: ${crearSilla().description()}\n Sofa: ${crearSofa().description()}\n Mesilla: ${crearMesilla().description()} "

}

class ModerFactory(): Factory(){

    override fun crearSilla(): Silla = SillaModerna()

    override fun crearSofa(): Sofa = SofaModerno()

    override fun crearMesilla(): Mesilla = MesillaModerna()

}

class VictorianFactory(): Factory(){

    override fun crearSilla(): Silla = SillaVictoriana()

    override fun crearSofa(): Sofa = SofaVictoriano()

    override fun crearMesilla(): Mesilla = MesillaVictoriana()

}

fun main(){
    entregarMuebles(ModerFactory())
    entregarMuebles(VictorianFactory())
}

fun entregarMuebles(factory: Factory){
    println(factory.combinarMuebles())
}

File: abstract_factory/dart/factory.dart
abstract class Silla {
  String description();
}

abstract class Sofa {
  String description();
}

abstract class Mesilla {
  String description();
}

// Tenemos diferente variadades de estos productos Moderno y Victoriano

//  Juego de estilo Victoriano
class SillaVictoriana extends Silla {
  @override
  String description() => 'silla victoriana';
}

class SofaVictoriano extends Sofa {
  @override
  String description() => 'sofa victoriano';
}

class MesillaVictoriana extends Mesilla {
  @override
  String description() => 'mesilla victoriano';
}

//Juego de estilo Moderno

class SillaModerna extends Silla {
  @override
  String description() => 'silla Moderna';
}

class SofaModerno extends Sofa {
  @override
  String description() => "sofa moderno";
}

class MesillaModerna extends Mesilla {
  @override
  String description() => "mesilla moderno";
}

abstract class Factory {
  Silla crearSilla();

  Sofa crearSofa();

  Mesilla crearMesilla();

  String combinarMuebles() =>
      "Juego de muebles Silla: ${crearSilla().description()}\n Sofa: ${crearSofa().description()}\n Mesilla: ${crearMesilla().description()} ";
}

class ModerFactory extends Factory {
  @override
  Mesilla crearMesilla() => MesillaModerna();

  @override
  Silla crearSilla() => SillaModerna();

  @override
  Sofa crearSofa() => SofaModerno();
}

class VictorianFactory extends Factory {
  @override
  Mesilla crearMesilla() => MesillaVictoriana();

  @override
  Silla crearSilla() => SillaVictoriana();

  @override
  Sofa crearSofa() => SofaVictoriano();
}

main() {
  entregarMuebles(ModerFactory());
  entregarMuebles(VictorianFactory());
}

entregarMuebles(Factory fact) {
  print(fact.combinarMuebles());
}
